[10.0] add new module to manage sale promotion based on rules

lmignon opened the pull request:
A new repository 'sale_discount' could be a better place for this kind of addon.

bguillot committed:
sale_promotion_rule: module to manage promotion rules

bguillot committed:
sale_promotion_rule: add button to clean promotion rule

lmignon committed:
sale_promotion_rule: Improve implementation
The module now supports multiple promotion rules.
When a coupon is used, it always take precedence on automatic promotion
rule.
The computation of promotions has been refactored to update lines from
the sale.order object. Sine lines are updated from the SO, the recompute
of SO fields that depends of lines is only tiggered at the end of the
update on the lines. (In place of each time a line is updated).
You can ask to recompute the promotion of a sale.order. If a coupon is
already specified it's conserved but all the automatic rules are applied
again.

lmignon committed:
[IMP] sale_promotion_rule: Remove demo data Avoid demo data with introducing a side effect on the expected results of all sale order created into unitest

lmignon committed:
[ADD] setup.py

sebastienbeau commented on sale_promotion_rule/__manifest__.py:
```diff
@@ -0,0 +1,33 @@
+    "version": "10.0.1.0.0",
+    "category": "Sale",
+    "website": "https://akretion.com",
+    "author": "Akretion",
```
Missing author "Acsone, Odoo Community Association (OCA)"
```suggestion
    "author": "Akretion, Acsone, Odoo Community Association (OCA)",
```

sebastienbeau commented on sale_promotion_rule/README.rst:
```diff
@@ -0,0 +1,44 @@
+* Add "automatic" rule that are automatically applied without coupon (only coupon have been implmented)
+* Add gift support. (Only discount offer have been implemented)
+* Add discount amount support
+* Improve condition by using custom filter on partner
```
Missing OCA README info

sebastienbeau reviewed (approved):
Fonctionnal tested
Thanks for proposing in OCA.
Minor change to do regarding author and readme

sebastienbeau commented on sale_promotion_rule/README.rst:
```diff
@@ -0,0 +1,44 @@
+
+* BEAU Sébastien <sebastien.beau@akretion.com>
+* Benoît GUILLOT <benoit.guillot@akretion.com>
+
```
you can add yourself (Laurent) as contributor

lmignon replied:
Thank you @sebastienbeau. Comments included in last commit.

lmignon committed:
[IMP] sale_promotion_rule: Improve README

legalsylvain commented on sale_promotion_rule/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,18 @@
+sale order. This rule takes always precedence over automatic rules.
+
+
+.. image:: https://raw.githubusercontent.com/OCA/sale-workflow/10.0/sale_promotion_rule/static/description/promotion_rule.png
```
You can now write : ../sale_promotion_rule/static/description/promotion_rule.png

(more simple, and make the url resilient, during the migration)

Not a blocking point.

lmignon replied:
thank you @legalsylvain Change done

rvalyi commented on sale_promotion_rule/__manifest__.py:
```diff
@@ -0,0 +1,34 @@
+    "development_status": "Beta",
+    "category": "Sale",
+    "website": "https://akretion.com",
+    'author': 'Anybox, ACSONE SA/NV,Odoo Community Association (OCA)',
```
this is strange: we have akretion in the website but not in the author

lmignon replied:
I'll fix it... Should be Akretion in place of Anybox...

lmignon replied:
@rvalyi fixed in https://github.com/OCA/sale-workflow/pull/729/commits/6fb82acd4c3538132d00a843acb58fe92afbb803

nikul-serpentcs commented on sale_promotion_rule/__manifest__.py:
```diff
@@ -0,0 +1,34 @@
+    "version": "10.0.1.0.0",
+    "development_status": "Beta",
+    "category": "Sale",
+    "website": "https://akretion.com",
```
Replace URL as per [this](https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py#L11)

lmignon replied:
@nikul-serpentcs done

nikul-serpentcs commented on sale_promotion_rule/__manifest__.py:
```diff
@@ -0,0 +1,34 @@
+    "license": "AGPL-3",
+    "application": False,
+    "installable": True,
+    "external_dependencies": {
```
@lmignon if not any `external_dependencies` then remove it

lmignon replied:
@nikul-serpentcs done

nikul-serpentcs reviewed (commented):
Improve Code

rousseldenis commented on sale_promotion_rule/models/sale_order.py:
```diff
@@ -0,0 +1,64 @@
+    def _compute_applied_promotion_rule_ids(self):
+        for rec in self:
+            rec.applied_promotion_rule_ids = (
+                rec.coupon_promotion_rule_id + rec.promotion_rule_ids)
```
Even if each rule is cumulated, IMHO use of union is better

lmignon replied:
done

rousseldenis commented on sale_promotion_rule/models/sale_promotion_rule.py:
```diff
@@ -0,0 +1,317 @@
+            ('no_restriction', 'No restriction')
+        ], default='no_restriction',
+        required=True)
+    restriction_amount = fields.Selection(
```
IMHO the name is quite weird as we are waiting for a float. Maybe 'restriction_amount_filed ?

lmignon replied:
done

rousseldenis commented on sale_promotion_rule/models/sale_promotion_rule.py:
```diff
@@ -0,0 +1,317 @@
+
+    def _check_valid_usage(self, order):
+        if self.usage_restriction == 'one_per_partner':
+            return not self.env['sale.order'].search([
```
Maybe a search_count is sufficent ?

lmignon replied:
done

rousseldenis commented on sale_promotion_rule/models/sale_promotion_rule.py:
```diff
@@ -0,0 +1,317 @@
+        required=True)
+    minimal_amount = fields.Float(
+        digits=dp.get_precision('Discount'))
+    display_name = fields.Char(
```
You should override name_get as display_name is automatically computed on it.

lmignon replied:
done

rousseldenis commented on sale_promotion_rule/models/sale_promotion_rule.py:
```diff
@@ -0,0 +1,317 @@
+                _('Discount promotion of type %s is not supported') %
+                self.discount_type)
+
+        lines_by_order = defaultdict(self.env['sale.order.line'].browse)
```
browse() ?

lmignon replied:
@rousseldenis arg of defaultdict must be a callable....

rousseldenis reviewed (changes requested):
LGTM. Some comments

lmignon committed:
[IMP] sale_promotion_rule: s/restriction_amount/restriction_amount_field
The field reference the name of the field to use on the sale order to determine if the promotion is applicable according to the total amount.

lmignon committed:
[IMP] sale_promotion_rule: Overrides name_get in place of defining a new field for display_name

lmignon committed:
[IMP] sale_promotion_rule: Improve performance
When checking if a promotion rule can be applied according to its usage definition, usage search_count in place of search since we are not interested by the resultset

pedrobaeza commented:
@lmignon may you squash a bit commit history before merging?

lmignon committed:
[FIX] sale_promotion_rule: We must also include Coupon into the multi rule strategy check

lmignon commented:
@pedrobaeza Sorry for the late reply, I didn't see your comment. It's a lot of work to rework/squash the commit history since I've 3 others PRs based on this one... 
https://github.com/OCA/sale-workflow/pulls/lmignon

pedrobaeza commented:
Well, tell me when you finish with all, and I'll merge this with a bit of squashing.

lmignon commented:
@pedrobaeza I've reworked the commit history. This PR is ready to merge. I'll rebase the others based on this one..
